bloom/consumed.py

```diff
@@ -0,0 +1,123 @@
+#-----------------------------------------------------------------------------#
+#   consumed.py                                                               #
+#                                                                             #
+#   Copyright (c) 2017-2018, Rajiv Bakulesh Shah, original author.            #
+#   All rights reserved.                                                      #
+#-----------------------------------------------------------------------------#
+
+
+import collections
+import json
+
+from .base import Base, run_doctests
+
+
+class RecentlyConsumed(Base):
+    '''Memcache-backed data structure to track recently consumed links.
+
+    On the consumer side, we want a data structure similar to a
+    collections.deque (as we're appending a user's new consumed links to the
+    right and popping old consumed links from the left).  But on the listing
+    service side, we want a data structure similar to a set (as we're testing
+    many candidate links for membership within the user's RecentlyConsumed).
+
+    Therefore, RecentlyConsumed (persisted in Memcache) implements a subset of
+    the collections.deque API and a subset of the set API.  As such, please
+    note that RecentlyConsumed silently ignores duplicate values during
+    append() and extend().
+
+    Further, please note that RecentlyConsumed is neither thread-safe nor safe
+    to use across instances.  Please lock around your RecentlyConsumed
+    instances/accesses.  A good choice for such a lock is .memlock.MemLock.
+    '''
+
+    _RANDOM_KEY_PREFIX = 'consumed:'
+    _MAXLEN = 1000
+
+    def __init__(self, memcache=None, key=None, maxlen=_MAXLEN):
+        super(RecentlyConsumed, self).__init__(memcache=memcache, key=key)
+        self._maxlen = maxlen
+        self._load_from_memcache()
+
+    @property
+    def maxlen(self):
+        return self._maxlen
+
+    @maxlen.setter
+    def maxlen(self, value):
+        message = "attribute 'maxlen' of '{}' objects is not writable"
+        message = message.format(self.__class__.__name__)
+        raise AttributeError(message)
+
+    def _load_from_memcache(self):
+        string = self.memcache.get(self.key, default='[]')
+        list_ = json.loads(string)
+        if self.maxlen is not None and len(list_) > self.maxlen:
+            message = 'persistent {} beyond its maximum size'
+            message = message.format(self.__class__.__name__)
+            raise IndexError(message)
+        self._deque = collections.deque(list_)
```

Comment:
Do the set and deque lazily instead?  We'll probably only ever be operating on one of them depending on whether we're reading or writing.